Add retry_if_exception_message

Brian-Williams opened the pull request:
From issue #125.

Add retry classes for exception messages.

I'm very open to name suggestions. Specifically wondering if `retry_if_not_exception_message` or `retry_until_exception_message` is best.

`retry_if_exception_message` and it's complement act different out of necessity. The normal class can end if there is no exception whereas it's inverse needs to retry if there is no exception message.

It's worth noting in my branch unit tests are intermittently failing on [wait class backwards incompatibility](https://travis-ci.org/Brian-Williams/tenacity/jobs/405236070#L470). I don't believe this is caused by any of my changes, but it's worth noting.

jd commented on tenacity/retry.py:
```diff
@@ -111,6 +112,61 @@ def __call__(self, attempt):
+            "{}() missing 1 required argument 'message' or 'match'".format(
+                self.__class__.__name__))
+
+    def _no_failure(self):
```
`@staticmethod`

Not sure it's worth having a method, just an attribute would be enough.

jd commented on tenacity/retry.py:
```diff
@@ -111,6 +112,61 @@ def __call__(self, attempt):
+
+        return inverted_predicate
+
+    def _no_failure(self):
```
@staticmethod

Not sure it's worth having a method, just an attribute would be enough.

jd commented on tenacity/retry.py:
```diff
@@ -111,6 +112,61 @@ def __call__(self, attempt):
+                "{}() takes either 'message' or 'match', not both".format(
+                    self.__class__.__name__))
+        self.predicate = self._predicate_factory(message, match)
+        self.require_exception = require_exception
```
That does not seem to be related to the heart of the function. I'd remove that, and put that in a different class (and PR) if that's needed.

jd commented on tenacity/retry.py:
```diff
@@ -111,6 +112,61 @@ def __call__(self, attempt):
             return not self.predicate(attempt.result())
 
 
+class retry_if_exception_message(retry_base):
```
you should inherit from `retry_if_exception`

jd commented on tenacity/retry.py:
```diff
@@ -111,6 +112,61 @@ def __call__(self, attempt):
+        """Retry logic for no failure."""
+        return False
+
+    def __call__(self, attempt):
```
This is not needed if you inherit from `retry_if_exception`

jd commented on tenacity/retry.py:
```diff
@@ -111,6 +112,61 @@ def __call__(self, attempt):
+            raise TypeError(
+                "{}() takes either 'message' or 'match', not both".format(
+                    self.__class__.__name__))
+        self.predicate = self._predicate_factory(message, match)
```
just pass the result to the `__init__` method of `retry_if_exception` here (since you will inherit from it). The factory method is not even really needed IMHO, you can put that code in `__init__`.

jd commented on tenacity/retry.py:
```diff
@@ -111,6 +112,61 @@ def __call__(self, attempt):
+
+class retry_if_not_exception_message(retry_if_exception_message):
+    """Retries until an exception message equals or matches."""
+
```
just set invert the predicate in `__init__` with a lambda I'd say.

Brian-Williams commented:
Thanks for the review!

Removed `require_exception`, which allowed us to inherit from `retry_if_exception` as suggested.

Removed `_no_failure` fnc as it no longer has a use case if `retry_if_not_exception_message` is the only `__call__` in this PR; `retry_if_exception_message` can simply inherit that magic method.

Fixed a bug in the assumption of the errormessage retrieval and wrote an additional unit test to handle the positive case of `retry_if_not_exception_message`.

Rebased on jd/master to a single commit.
